src/reconcile/rec_row.c

```diff
@@ -467,25 +466,42 @@ __wt_rec_row_int(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_PAGE *page)
             key->cell_len = 0;
             key->len = key->buf.size;
             ovfl_key = true;
+            page_image += ikey->size;
         } else {
             __wt_ref_key(page, ref, &p, &size);
-            WT_ERR(__rec_cell_build_int_key(session, r, p, r->cell_zero ? 1 : size, &ovfl_key));
+            if (r->cell_zero)
+                size = 1;
+            WT_ERR(__rec_cell_build_int_key(session, r, p, size, &ovfl_key));
+            if (ovfl_key)
+                page_image += size;
         }
         r->cell_zero = false;
 
-        /* Boundary: split or write the page. */
-        if (__wt_rec_need_split(r, key->len + val->len)) {
+        /*
+         * Boundary: split or write the page.
+         *
+         * We always instantiate row-store internal page keys in order to avoid special casing the
+         * B+tree search code to handle keys that may not exist (and I/O in a search path). Because
+         * root pages are pinned, overflow keys on the root page can cause follow-on cache effects
+         * where huge root pages tie down lots of cache space and eviction frantically attempts to
+         * evict objects that can't be evicted. If the in-memory image is too large, force a split.
+         * Potentially, limiting ourselves to 10 items per page is going to result in deep trees
+         * which will impact search performance, but at some point, the application's configuration
+         * is too stupid to survive.
+         */
+        force = r->entries > 20 && page_image > btree->maxmempage_image;
```

Comment:
Is 10 per page a magic figure here? I do not have the knowledge to review how good or bad "10" would behave in this case.

Reply:
Me either. Basically, at this point we've created an internal page that will result in a too-large page in memory, so it's a trade-off between too few keys in the internal page and too large a page in memory.  I'm not sure what we can do other than bound the page to some combination of the above.  I don't think there are any good answers.